test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java

```diff
@@ -0,0 +1,126 @@
+import org.testng.annotations.Test;
+
+import java.io.*;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipException;
+import java.util.zip.ZipFile;
+import java.util.zip.ZipOutputStream;
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.fail;
+
+public class CenSizeTooLarge {
+
+    private static final int ENDHDR = 22;          // End of central directory record size
+    private static final int CEN_SIZE_OFFSET = 12; // Offset of CEN size field within ENDHDR
+    private Path huge; // ZIP file with CEN size exceeding limit
+    private Path big;  // ZIP file with CEN size exactly on the limit
+
+    /**
+     * Create ZIP files with CEN sizes exactly on and exceeding the CEN size limit
+     */
+    @BeforeTest
+    public void setup() throws IOException {
+        var limit = Integer.MAX_VALUE - ENDHDR - 1;
+        var exceedingLimit = limit + 1;
+        big = zipWithCenSize("bigZip.zip", limit);
+        huge = zipWithCenSize("hugeZip.zip", exceedingLimit);
+    }
+
+    /**
+     * Validates that an end of central directory record with
+     * a CEN length exceeding the CEN limit is rejected
+     */
+    @Test
+    public void shouldRejectTooLargeCenSize() throws IOException {
+        assertRejected(huge, "invalid END header (central directory size too large)");
+    }
+
+    /**
+     * Validate that an end of central directory record with a
+     * valid CEN size is not rejected because of its CEN size
+     */
+    @Test
+    public void shouldRejectInvalidCenSize() throws IOException {
+        // Since the file has just a single entry, the CEN size in the END
+        // record doesn not match the real CEN and file size.
+        // Expect the CEN size should be rejected as invalid
+        assertRejected(big, "invalid END header (bad central directory size)");
+    }
+
+    /**
+     * Assert that opening a file with ZipFile throws with a ZipException
+     * with the expected message
+     */
+    private void assertRejected(Path zip, String expectedMsg) throws IOException {
+        try (ZipFile zf = new ZipFile(zip.toFile())) {
+            fail("Expected ZipFile to throw ZipException");
+        } catch (ZipException e) {
+            var actual = e.getMessage();
+            assertTrue(expectedMsg.equals(actual),
+                    "Expected ZipException message '%s', got '%s'".formatted(expectedMsg, actual));
+        }
+    }
+
+    /**
+     * Create an ZIP file with a single entry, then modify the CEN size
+     * in the End of central directory record to the given size
+     *
+     * The resulting ZIP is technically not valid, but it does allow us
+     * to test that the large CEN size is rejected
+     */
+    private Path zipWithCenSize(String name, int cenSize) throws IOException {
```

Comment:
You could speed this up a bit more by just creating the array representing the original Zip 1 time and modify it each pass.(though it might not speed things up that much)

Reply:
I introduced the templateZip() method which returns a byte[]. While the speed is insignificat here, I think it makes it more explicit that we're producing ZIPs from a template.